mesonbuild/dependencies.py

```diff
@@ -1599,6 +1600,117 @@ def __init__(self, environment, kwargs):
     def get_link_args(self):
         return []
 
+class LLVMDependency(Dependency):
+    """LLVM dependency.
+
+    LLVM uses a special tool, llvm-config, which has arguments for getting
+    c args, cxx args, and ldargs as well as version.
+    """
+
+    # Ordered list of llvm-config binaries to try. Start with base, then try
+    # newest back to oldest (3.5 is abitrary), and finally the devel version.
+    llvm_config_bins = [
+        'llvm-config', 'llvm-config-4.0', 'llvm-config-3.9', 'llvm-config39',
+        'llvm-config-3.8', 'llvm-config38', 'llvm-config-3.7', 'llvm-config37',
+        'llvm-config-3.6', 'llvm-config36', 'llvm-config-3.5', 'llvm-config35',
+        'llvm-config-3.4', 'llvm-config34', 'llvm-config-3.3', 'llvm-config33',
+        'llvm-config-devel',
+    ]
+    llvmconfig = None
+    _llvmconfig_found = False
+    __best_found = None
+
+    def __init__(self, environment, kwargs):
+        super().__init__('llvm-config', kwargs)
+        self.cargs = []
+        self.libs = []
+        self.components = []
+
+        req_version = kwargs.get('version', None)
+        if self.llvmconfig is None:
+            self.check_llvmconfig(req_version)
+        if not self._llvmconfig_found:
+            if self.__best_found is not None:
+                mlog.log('found {!r} but need:'.format(self.version),
+                         req_version)
+            else:
+                mlog.log("No llvm-config found; can't detect dependency")
+            mlog.log('Dependency LLVM found:', mlog.red('NO'))
+            if kwargs.get('required', True):
+                raise DependencyException('Dependency LLVM not found')
+            return
+
+        p, out = Popen_safe([self.llvmconfig, '--version'])[:2]
+        if p.returncode != 0:
+            mlog.log('Dependency LLVM found:', mlog.red('NO'))
+        else:
+            self.version = out.strip()
+            mlog.log('Dependency LLVM found:', mlog.green('YES'))
+            self.is_found = True
+
+            p, out = Popen_safe([self.llvmconfig, '--libs'])[:2]
+            if p.returncode != 0:
+                raise DependencyException('Could not generate libs for LLVM.')
```

Comment:
I think this and all later `DependencyException`s should only be thrown if `required : false`. Otherwise, a useful warning message should be printed before printing `found: NO`.

It's good to always error out on invalid syntax or input in the meson build files, but not for things that are outside the user's control, such as a broken llvm installation.